## DSML investigation

You are part of the Suisse Impossible Mission Force, or SIMF for short. You need to uncover a rogue agent that is trying to steal sensitive information.

Your mission, should you choose to accept it, is to find that agent before stealing any classified information. Good luck!

# Assignment part two

We received new intelligence about the rogue agent which lets us discover his age.
Use the table "social_media_SIMF_archive.csv" to create a model that predicts the age of a person. Use this model to predict the ages of the table of suspects "suspect_social_media_statistics.csv".



## Getting to know our data

- userId: Unique number given to each user.

- tier: Tier of the city in which the user is residing.

- gender: Categorical feature representing the gender of the user. 1 represents male and 2 represents female.

- following_rate: Number of accounts followed by the user(feature is normalized)

- followers_avg_age: Average of age groups of all the followers of the user.

- following_avg_age: Average of age groups of all the accounts followed by the user.

- max_repetitive_punc: Maximum repititive punctuations found in the bio and comments of the user.

- num_of_hashtags_per_action: Average nubmer of hashtags used by the user per comment.

- emoji_count_per_action: Average number of emojis used by the user per comment

- punctuations_per_action: Average number of punctuations used by the user per comment.

- number_of_words_per_action: Average number of words used by the user per comment.

- avgCompletion: Average watch time completion rate of the videos.

- avgTimeSpent: Average time spent by the user on a video in seconds.

- avgDuration: Average duration of the videos that the user has watched till date.

- avgComments: Average number of comments per video watched.

- creations: Total number of videos uploaded by the user.

- content_views: Total number of videos watched.

- num_of_comments: Total number of comments made by the user (normalized)

- weekends_trails_watched_per_day: Number of videos watched on weekends per day.

- weekdays_trails_watched_per_day: Number of videos watched on weekdays per day.

- slot1_trails_watched_per_day: The day is divided into 4 slots. This feature represents the average number of videos watched in this particular time slot.

- slot2_trails_watched_per_day: The day is divided into 4 slots. This feature represents the average number of videos watched in this particular time slot.

- slot3_trails_watched_per_day: The day is divided into 4 slots. This feature represents the average number of videos watched in this particular time slot.

- slot4_trails_watched_per_day: data

- avgt2: Average number of followers of all the accounts followed by the user.

- age : The age of the user


In [225]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
train_url = pd.read_csv("https://raw.githubusercontent.com/michalis0/DataScience_and_MachineLearning/refs/heads/master/Assignements/Part%202/data/social_media_SIMF_archive.csv")
suspects_url = pd.read_csv("https://raw.githubusercontent.com/michalis0/DataScience_and_MachineLearning/refs/heads/master/Assignements/Part%202/data/suspects_social_media.csv")

In [226]:
np.random.seed(42)

### Training a model
Select the features which by convention are called X. We will also choose the target variable which we typically call y.

You can use the features `'tier', 'gender', 'following_rate', 'followers_avg_age', 'following_avg_age', 'max_repetitive_punc', 'num_of_hashtags_per_action', 'emoji_count_per_action', 'punctuations_per_action', 'number_of_words_per_action', 'avgCompletion', 'avgTimeSpent', 'avgDuration', 'avgComments', 'creations', 'content_views', 'num_of_comments', 'weekends_trails_watched_per_day', 'weekdays_trails_watched_per_day', 'slot1_trails_watched_per_day', 'slot2_trails_watched_per_day', 'slot3_trails_watched_per_day', 'slot4_trails_watched_per_day', 'avgt2'`


In [227]:
train = train_url.drop_duplicates().dropna()  # drop duplicates and NaN values
display(train.head())                         # Display a sample of the data
print(train.dtypes)                           # Print the data types
features = train.loc[:,['age','tier', 'gender', 'following_rate', 'followers_avg_age', 'following_avg_age', 'max_repetitive_punc', 'num_of_hashtags_per_action', 'emoji_count_per_action', 'punctuations_per_action', 'number_of_words_per_action', 'avgCompletion', 'avgTimeSpent', 'avgDuration', 'avgComments', 'creations', 'content_views', 'num_of_comments', 'weekends_trails_watched_per_day', 'weekdays_trails_watched_per_day', 'slot1_trails_watched_per_day', 'slot2_trails_watched_per_day', 'slot3_trails_watched_per_day', 'slot4_trails_watched_per_day', 'avgt2']]
display(features.describe())                  # Summary statistics
display(features.corr(numeric_only = True))   # Correlation matrix

Unnamed: 0  Unnamed: 0.1    userId  tier  gender  following_rate  \
0           0             0  47496533     2       1        0.000000   
1           1             1    584405     3       1        0.011096   
2           2             2  39558663     2       2        0.064516   
3           3             3  35027479     2       1        0.000000   
4           4             4  59099484     2       1        0.357143   

   followers_avg_age  following_avg_age  max_repetitive_punc  \
0                0.0           0.000000                    0   
1                3.0           3.000000                    0   
2                1.4           2.142857                    0   
3                0.0           0.000000                    0   
4                3.0           2.000000                    7   

   num_of_hashtags_per_action  ...  content_views  num_of_comments  \
0                         0.0  ...       0.190083         0.000000   
1                         0.0  ...       0.020804         0.000000   
2                         0.0  ...       0.104839         0.000000   
3                         0.0  ...       0.007937         0.000000   
4                         0.0  ...       0.765306         0.030612   

   weekends_trails_watched_per_day  weekdays_trails_watched_per_day  \
0                         0.000000                         0.039669   
1                         0.004854                         0.003329   
2                         0.060484                         0.001613   
3                         0.000000                         0.001587   
4                         0.000000                         0.163265   

   slot1_trails_watched_per_day  slot2_trails_watched_per_day  \
0                      0.008264                      0.000000   
1                      0.004161                      0.004161   
2                      0.104839                      0.016129   
3                      0.000000                      0.007937   
4                      0.000000                      0.204082   

   slot3_trails_watched_per_day  slot4_trails_watched_per_day        avgt2  \
0                      0.132231                      0.057851     0.000000   
1                      0.005548                      0.012483  6709.600000   
2                      0.008065                      0.000000   138.857143   
3                      0.000000                      0.000000     0.000000   
4                      0.612245                      0.000000   177.500000   

         age  
0  38.599100  
1  77.807570  
2  61.736974  
3  37.066390  
4  69.419177  

[5 rows x 28 columns]

Unnamed: 0                           int64
Unnamed: 0.1                         int64
userId                               int64
tier                                 int64
gender                               int64
following_rate                     float64
followers_avg_age                  float64
following_avg_age                  float64
max_repetitive_punc                  int64
num_of_hashtags_per_action         float64
emoji_count_per_action             float64
punctuations_per_action            float64
number_of_words_per_action         float64
avgCompletion                      float64
avgTimeSpent                       float64
avgDuration                        float64
avgComments                          int64
creations                          float64
content_views                      float64
num_of_comments                    float64
weekends_trails_watched_per_day    float64
weekdays_trails_watched_per_day    float64
slot1_trails_watched_per_day       float64
slot2_trail

age           tier         gender  following_rate  \
count  486049.000000  486049.000000  486049.000000   486049.000000   
mean       45.710512       1.975930       1.213005        0.076561   
std        12.064161       0.430629       0.409432        2.238261   
min        13.982634       1.000000       1.000000        0.000000   
25%        37.270653       2.000000       1.000000        0.000000   
50%        41.110277       2.000000       1.000000        0.000000   
75%        52.671177       2.000000       1.000000        0.008547   
max        94.998682       3.000000       2.000000      895.304000   

       followers_avg_age  following_avg_age  max_repetitive_punc  \
count      486049.000000      486049.000000        486049.000000   
mean            0.341680           0.396869             0.706293   
std             0.808689           0.918341             2.508147   
min             0.000000           0.000000             0.000000   
25%             0.000000           0.000000             0.000000   
50%             0.000000           0.000000             0.000000   
75%             0.000000           0.000000             0.000000   
max             4.000000           4.000000            75.000000   

       num_of_hashtags_per_action  emoji_count_per_action  \
count               486049.000000           486049.000000   
mean                     0.000279                0.000981   
std                      0.012256                0.020232   
min                      0.000000                0.000000   
25%                      0.000000                0.000000   
50%                      0.000000                0.000000   
75%                      0.000000                0.000000   
max                      2.333333                3.000000   

       punctuations_per_action  ...      creations  content_views  \
count            486049.000000  ...  486049.000000  486049.000000   
mean                  0.012714  ...       0.014856       0.373715   
std                   0.154763  ...       0.067143       0.759658   
min                   0.000000  ...       0.000000       0.000887   
25%                   0.000000  ...       0.000000       0.040816   
50%                   0.000000  ...       0.000000       0.123967   
75%                   0.000000  ...       0.009091       0.362205   
max                  24.000000  ...       9.571429      17.150000   

       num_of_comments  weekends_trails_watched_per_day  \
count    486049.000000                    486049.000000   
mean          0.001756                         0.070666   
std           0.019448                         0.179313   
min           0.000000                         0.000000   
25%           0.000000                         0.000000   
50%           0.000000                         0.003968   
75%           0.000000                         0.059524   
max           4.340611                         6.014563   

       weekdays_trails_watched_per_day  slot1_trails_watched_per_day  \
count                    486049.000000                 486049.000000   
mean                          0.063329                      0.030087   
std                           0.141353                      0.147136   
min                           0.000000                      0.000000   
25%                           0.002273                      0.000000   
50%                           0.015800                      0.000000   
75%                           0.058621                      0.000000   
max                           2.960345                      8.008000   

       slot2_trails_watched_per_day  slot3_trails_watched_per_day  \
count                 486049.000000                 486049.000000   
mean                       0.136383                      0.142864   
std                        0.354379                      0.378985   
min                        0.000000                      0.000000   
25%                        0.000000                      0.000000   
50%   

age      tier    gender  following_rate  \
age                              1.000000 -0.165830 -0.091494        0.040759   
tier                            -0.165830  1.000000 -0.016361       -0.017286   
gender                          -0.091494 -0.016361  1.000000       -0.007468   
following_rate                   0.040759 -0.017286 -0.007468        1.000000   
followers_avg_age                0.648654 -0.082364 -0.031083        0.066589   
following_avg_age                0.651393 -0.080192 -0.039052        0.062658   
max_repetitive_punc              0.326175 -0.062463  0.031989        0.038079   
num_of_hashtags_per_action       0.011309 -0.003512 -0.000566       -0.000237   
emoji_count_per_action           0.009719  0.000023  0.016944        0.000014   
punctuations_per_action          0.045945 -0.002348  0.008759        0.000714   
number_of_words_per_action       0.358135 -0.072673 -0.001973        0.016876   
avgCompletion                    0.137776 -0.005351  0.059360       -0.001335   
avgTimeSpent                     0.003670 -0.003233  0.002640       -0.000042   
avgDuration                      0.100453 -0.011028 -0.026138        0.002714   
avgComments                      0.107172 -0.047288  0.009702        0.029510   
creations                        0.257104 -0.086137  0.016714        0.084310   
content_views                    0.302707 -0.074116  0.045277        0.045564   
num_of_comments                  0.127609 -0.041715  0.007687        0.048179   
weekends_trails_watched_per_day  0.277706 -0.066890  0.043831        0.049063   
weekdays_trails_watched_per_day  0.310725 -0.084095  0.045265        0.054572   
slot1_trails_watched_per_day     0.179463 -0.043495  0.013383        0.027134   
slot2_trails_watched_per_day     0.227908 -0.059739  0.020152        0.047532   
slot3_trails_watched_per_day     0.259649 -0.072063  0.065264        0.044431   
slot4_trails_watched_per_day     0.287310 -0.072172  0.035060        0.046750   
avgt2                            0.283203 -0.040861 -0.020380        0.001832   

                                 followers_avg_age  following_avg_age  \
age                                       0.648654           0.651393   
tier                                     -0.082364          -0.080192   
gender                                   -0.031083          -0.039052   
following_rate                            0.066589           0.062658   
followers_avg_age                         1.000000           0.939457   
following_avg_age                         0.939457           1.000000   
max_repetitive_punc                       0.189000           0.180762   
num_of_hashtags_per_action                0.007385           0.007621   
emoji_count_per_action                    0.003327           0.003535   
punctuations_per_action                   0.012788           0.013128   
number_of_words_per_action                0.162000           0.164295   
avgCompletion                            -0.003958          -0.007674   
avgTimeSpent                              0.003276           0.003174   
avgDuration                               0.071725           0.077301   
avgComments                               0.101380           0.099449   
creations                                 0.195814           0.183056   
content_views                             0.125641           0.110422   
num_of_comments                           0.116840           0.111314   
weekends_trails_watched_per_day           0.139231           0.126204   
weekdays_trails_watched_per_day           0.147099           0.130550   
slot1_trails_watched_per_day              0.087722           0.079925   
slot2_trails_watched_per_day              0.118190           0.105560   
slot3_trails_watched_per_day              0.123388           0.109352   
slot4_trails_watched_per_day              0.132350           0.118631   
avgt2                                     0.114741           0.120905   

                                

Now do the train-test split, using 80% of the data as training and 20% as test. Set the random state to 42.

In [228]:
independent = features[['age']]
dependent = features.drop(columns=['age'])
dependent_train, dependent_test, independent_train, independent_test = train_test_split(dependent, independent, test_size=0.2, random_state=42, shuffle=True)

Next, rescale the features in training and test data using `MinMaxScaler`.

In [229]:
scaler = MinMaxScaler()
dependent_train = scaler.fit_transform(dependent_train)
dependent_test = scaler.transform(dependent_test)

## Linear regression
Now, create a linear regression model. Train it on the training data and use the model to make predictions on the test data. How is the performance of your model on the test data?

*Note: For the liner regression model, use the default attribute value as seen in the Lab.*

In [230]:
model = LinearRegression()

model.fit(dependent_train, independent_train)

print("R\u00b2 Score of the model: ",
      round(model.score(dependent_train, independent_train), 3))

print("Intercept: ", model.intercept_[0])
model_coeff = pd.DataFrame(model.coef_.flatten(),
                     index=['tier', 'gender', 'following_rate', 'followers_avg_age', 'following_avg_age', 'max_repetitive_punc', 'num_of_hashtags_per_action', 'emoji_count_per_action', 'punctuations_per_action', 'number_of_words_per_action', 'avgCompletion', 'avgTimeSpent', 'avgDuration', 'avgComments', 'creations', 'content_views', 'num_of_comments', 'weekends_trails_watched_per_day', 'weekdays_trails_watched_per_day', 'slot1_trails_watched_per_day', 'slot2_trails_watched_per_day', 'slot3_trails_watched_per_day', 'slot4_trails_watched_per_day', 'avgt2'],
                     columns=['Coefficients multivariate model'])
model_coeff

R² Score of the model:  0.661
Intercept:  36.657591731244246


Coefficients multivariate model
tier                                                   -3.929323
gender                                                 -2.750156
following_rate                                        -54.607097
followers_avg_age                                      12.952534
following_avg_age                                      16.481645
max_repetitive_punc                                    54.958655
num_of_hashtags_per_action                            -18.823607
emoji_count_per_action                                -30.077548
punctuations_per_action                               -47.425984
number_of_words_per_action                             96.322587
avgCompletion                                          10.963554
avgTimeSpent                                            0.485035
avgDuration                                           155.800963
avgComments                                           -33.621575
creations                                              -6.347380
content_views                                          24.332658
num_of_comments                                        -4.737368
weekends_trails_watched_per_day                      2650.314593
weekdays_trails_watched_per_day                      3014.038566
slot1_trails_watched_per_day                        -1719.554227
slot2_trails_watched_per_day                        -2480.163410
slot3_trails_watched_per_day                        -2616.671252
slot4_trails_watched_per_day                        -2649.785702
avgt2                                                  70.840337

and we do the predictions on the test set:

In [231]:
predictions = model.predict(dependent_test)

What are the model's $MAE$, $MSE$ and $R^2$ ?

In [232]:
mae = mean_absolute_error(independent_test, predictions)
mse = mean_squared_error(independent_test, predictions)
r2 = r2_score(independent_test, predictions)

print(f"MAE test set: {mae:0.2f};")
print(f"MSE test set: {mse:0.2f};")
print(f"R\u00b2 test set: {r2:0.2f};" )

MAE test set: 5.33;
MSE test set: 48.95;
R² test set: 0.66;


**Q1. What is the mean absolute error (MAE) of the model over the test set?**

*Hint : The correct value should be between 5.20 and 5.50*

`5.33`

**Q2. What is the mean squared error (MSE) of the model over the test set?"**

*Hint : The correct value should be between 48 and 50*

`48.95`

**Q3. What is the $R^2$ score of the model over the test set?**

*Hint : The correct value should be between 0.55 and 0.75*

`0.66`

## Retraining on the full dataset
Training on a split of the training dataset, as you just did, is mainly for evaluating the performance of the resulting model on the unseen data and also optimizing various hyper-parameters of the model.

We can now retrain the model on the full training data prior to making predictions.

Before predicting the age of the suspects, retrain the model on the full training dataset.

You can reuse the same features as before, but make sure those are also available on the suspects dataset.

In [233]:
train_df = pd.read_csv("https://raw.githubusercontent.com/michalis0/DataScience_and_MachineLearning/refs/heads/master/Assignements/Part%202/data/social_media_SIMF_archive.csv")
train_clean = train_df.drop_duplicates().dropna()

feature_cols = [
    'tier', 'gender', 'following_rate', 'followers_avg_age',
    'following_avg_age', 'max_repetitive_punc', 'num_of_hashtags_per_action',
    'emoji_count_per_action', 'punctuations_per_action', 'number_of_words_per_action',
    'avgCompletion', 'avgTimeSpent', 'avgDuration', 'avgComments', 'creations',
    'content_views', 'num_of_comments', 'weekends_trails_watched_per_day',
    'weekdays_trails_watched_per_day', 'slot1_trails_watched_per_day',
    'slot2_trails_watched_per_day', 'slot3_trails_watched_per_day',
    'slot4_trails_watched_per_day', 'avgt2']

X_full_train = train_clean[feature_cols]
y_full_train = train_clean['age']

In [234]:
full_scaler = MinMaxScaler()
X_full_train_scaled = full_scaler.fit_transform(X_full_train)

In [235]:
final_model = LinearRegression()
final_model.fit(X_full_train_scaled, y_full_train)

LinearRegression()

In [236]:
suspects_df = pd.read_csv("https://raw.githubusercontent.com/michalis0/DataScience_and_MachineLearning/refs/heads/master/Assignements/Part%202/data/suspects_social_media.csv")

X_suspects = suspects_df[feature_cols]
X_suspects_scaled = full_scaler.transform(X_suspects)

predicted_ages = final_model.predict(X_suspects_scaled)
suspects_df['predicted_age'] = predicted_ages

print("Suspects' ages predicted successfully!")
display(suspects_df[['userId', 'predicted_age']].head())

Suspects' ages predicted successfully!


userId  predicted_age
0  317991      38.483885
1  241892      36.929202
2  303376      54.806991
3  761992      53.171815
4  373318      46.475667

**Keep only the suspects between 25 and 35 years old. Use the `age` values you predicted in the previous step.**


In [237]:
filtered_suspects = suspects_df[(suspects_df['predicted_age'] >= 25) & (suspects_df['predicted_age'] <= 35)]
suspect_ids_to_check = [410319, 785994, 123699, 757865, 255830]
remaining_suspects_df = filtered_suspects[filtered_suspects['userId'].isin(suspect_ids_to_check)]

print(remaining_suspects_df['userId'].tolist())

[785994, 123699, 255830, 410319]


**Q4. Which of these users remain suspects after filtering for those aged between 25 and 35?** (check all that apply)

`785994`, `123699`, `255830`, `410319`

The SIMF is considering narrowing down the suspect pool by focusing on gender. As a preliminary step, they need to know the percentage of female suspects within the age range of 25 to 35 years old.

Calculate and report the percentage of female suspects among those aged 25 to 35 years old from the suspects you identified previously.

*Hint: In the provided dataset, 'gender' is coded with 1 representing male users and 2 representing female users.*

In [238]:
total_filtered_suspects = len(filtered_suspects)

female_suspects = len(filtered_suspects[filtered_suspects['gender'] == 2])

if total_filtered_suspects > 0:
    percentage_female = (female_suspects / total_filtered_suspects) * 100
else:
    percentage_female = 0

print(f"Total suspects aged 25-35: {total_filtered_suspects}")
print(f"Female suspects aged 25-35: {female_suspects}")
print(f"Percentage of female users: {percentage_female:.2f}%")

Total suspects aged 25-35: 22
Female suspects aged 25-35: 21
Percentage of female users: 95.45%


**Q5. What is the percentage of female users among the predicted suspects aged between 25 and 35 years old?**

`95.45`



**What is the predicted age of a Female user with the following characteristics:**
* tier = 2
* following_rate = 2
* followers_avg_age = 27
* following_avg_age = 22
* max_repetitive_punc = 0.06
* num_of_hashtags_per_action = 0.8
* emoji_count_per_action = 0.3
* punctuations_per_action = 0.03
* number_of_words_per_action = 7
* avgCompletion = 12
* avgTimeSpent = 0.13
* avgDuration = 0.8
* avgComments = 0.04
* likeToDislikeRation = 0.76
* creations = 10
* content_views = 65
* num_of_comments = 12
* weekends_trails_watched_per_day = 0
* weekdays_trails_watched_per_day = 0.00353
* slot1_trails_watched_per_day = 0.032
* slot2_trails_watched_per_day = 0
* slot3_trails_watched_per_day = 0.004
* slot4_trails_watched_per_day = 2
* avgt2 = 39

*hint: make sure the features and their order corresponds to the ones used for the training.*

*hint 2: don't forget to reprocess the data as you did before*

In [239]:
new_user_data = {
    'tier': 2,
    'gender': 2,  # Female
    'following_rate': 2,
    'followers_avg_age': 27,
    'following_avg_age': 22,
    'max_repetitive_punc': 0.06,
    'num_of_hashtags_per_action': 0.8,
    'emoji_count_per_action': 0.3,
    'punctuations_per_action': 0.03,
    'number_of_words_per_action': 7,
    'avgCompletion': 12,
    'avgTimeSpent': 0.13,
    'avgDuration': 0.8,
    'avgComments': 0.04,
    'creations': 10,
    'content_views': 65,
    'num_of_comments': 12,
    'weekends_trails_watched_per_day': 0,
    'weekdays_trails_watched_per_day': 0.00353,
    'slot1_trails_watched_per_day': 0.032,
    'slot2_trails_watched_per_day': 0,
    'slot3_trails_watched_per_day': 0.004,
    'slot4_trails_watched_per_day': 2,
    'avgt2': 39
}

new_user_df = pd.DataFrame([new_user_data], columns=feature_cols)
new_user_scaled = full_scaler.transform(new_user_df)
predicted_age_new_user = final_model.predict(new_user_scaled)
final_age = round(predicted_age_new_user[0])

print(f"Predicted Age (raw): {predicted_age_new_user[0]}")
print(f"Predicted Age (rounded): {final_age}")

Predicted Age (raw): 26.25457360883418
Predicted Age (rounded): 26


**Q6. What is the predicted age of a female user with those characteristics using the regression model that you created?**

`26`

## Your investigation is progressing effectively, and the list of suspects is narrowing down.

**Don't forget to answer the quiz and submit your code on Moodle before the end of the deadline.**